### DISCUSSION TITLE
User-defined functions?

### DISCUSSION THREAD

**Discussion by zombiezen** [cuelang/cue#484] [General] [answered]
Is there a convenient way of creating user-defined functions or something similar? Under [boilerplate removal](https://cuelang.org/docs/concepts/logic/#boilerplate-removal), the `pattern constraint` is used to imply a default constraint, but sometimes I want to build a value based on some set of input parameters (not just a single string). The best I can come up with now is:

```
#udf: {
  #foo: string
  out: "\(#foo)bar"
}

"quux": (#udf & {#foo: "foo"}).out
```

**Comment by zombiezen** [comment]
From @myitcv in [Slack](https://cuelang.slack.com/archives/CLT4FD7KP/p1597347652120000): this is the correct way to go for now, but the new evaluator will make this better by allowing non-structs to be emitted. Combined with type cast semantics (like those I proposed in #454), this will make these sorts of limited functions much easier to work with.

**Reply by mpvl** [reply]
We are also contemplating a syntactic sugar for the arguments, allowing the above to be written as
```
#udf: {
  #0: string
  
  "\(#0)bar"
}

"quux": #udf("foo")
```

`#0` would just be a regular field otherwise. This approach avoids some of the pitfalls of having functions in a configuration language.

Note that the new evaluator now explicitly disallows reentrancy. This was already disallowed by the spec, but not previously enforced.

**Reply by mpvl** [reply]
For completeness, we are also investigating a native bridge to some regular language, perhaps only to be done in the tooling layer.

The idea is that a configuration language is not the best place to do computation and that this is best left to an established general programming language. Rather than enhancing CUE with the ability of computation, we use CUE's commutativity to mix in computed results. The strict separation of computation and configuration, combined with CUE's commutativity property, allows computation to be ignored leaving a non-concrete pure CUE configuration that can be refactored and analyzed with tools like trim and anti-unification. Many of the possible analysis tools would break down if we start to blur the lines between computation and configuration.

Either way, the idea is still to allow simple "macros" to allow dealing with simple cases within CUE itself.

**Reply by myitcv** [reply]
For those who view this answer, there was further discussion about the syntactic sugar for functions in https://github.com/cuelang/cue/issues/69#issuecomment-739228231

**Reply by PierreR** [reply]
Talking about `native bridge` is webAssembly in the picture to propose an interface for various programming languages ?

**Comment by myitcv** [comment, marked as answer]
The current best answer is:

```
foo: {#a: int, #b: int, #a + #b } // struct with embedded scalar
call: foo&{_, #a: 1, #b: 2}         // 3
```

(a change from the previously marked answer because we now have full support for embedded scalars)

There is a proposal to support a syntactic sugar based on this:

```
foo: { #a: int, #b: int, #a + #b } // struct with embedded scalar
call: foo(#a: 1, #b: 2) // 3
```

**Reply by codewithcheese** [reply]
@myitcv with `cue eval --concrete` the above returns `foo: non-concrete value int in operand to +:
    ./demo/func.cue:25:25`

**Reply by myitcv** [reply]
@codewithcheese please can you raise an issue if you're seeing an issue with this? Thanks

**Reply by morlay** [reply]
@myitcv your case will be throw the errors.  (v0.3.0-beta.7)

@codewithcheese  function should be a hidden one `_` or `#` prefix

```
_foo: {#a: int, #b: int, #a + #b }
call: _foo & {_, #a: 1, #b: 2} 
```

**Reply by myitcv** [reply]
@morlay 

> @myitcv your case will be throw the errors. (v0.3.0-beta.7)

Apologies, I'm not sure what you're referring to here. Are you suggesting there's an issue with `beta.7`? If so, please can you give an example of what is failing?

**Comment by vikstrous2** [comment]
Taking this idea a step further and adding a conditional:
```
cond: true
foo: {
  #a: int
  #b: int
  if cond  { output: #a + #b }
  if !cond { output: 4 }
} // struct with embedded scalar
call: (foo&{#a: 1, #b: 2}).output         // 3
```
The conditional version doesn't work with the embedding scalars approach.

**Reply by myitcv** [reply]
It should work with embedded scalars. So I just raised https://github.com/cuelang/cue/issues/729. Thanks for pointing this out.

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/484.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
